src/coreclr/jit/lowerxarch.cpp

```diff
@@ -320,9 +320,223 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
+        {
+            GenTree* next;
+            if (TryLowerAndOrToCCMP(binOp, &next))
+            {
+                return next;
+            }
+        }
+    }
+#endif // TARGET_AMD64
+
     return binOp->gtNext;
 }
 
+#ifdef TARGET_AMD64
+
+//------------------------------------------------------------------------
+// ProducesPotentialConsumableFlagsForCCMP : Checks if an op will set
+// the condition flags in a form that ccmp may us (either an integer relop
+// or a setcc).
+//
+// Arguments:
+//    op - tree to check for consumable flags
+//
+// Return Value:
+//    true if the op output in form of flags can be used by ccmp
+//
+bool Lowering::ProducesPotentialConsumableFlagsForCCMP(GenTree* op)
+{
+    if (op->OperIsCompare())
+    {
+        GenTreeOp* relop = op->AsOp();
+        return !GenCondition::FromRelop(relop).IsFloat();
+    }
+
+    if (op->OperIs(GT_SETCC))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------
+// TryLowerAndOrToCCMP : Lower AND/OR of two conditions into test + CCMP + SETCC nodes.
+//
+// Arguments:
+//    tree - pointer to the node
+//    next - [out] Next node to lower if this function returns true
+//
+// Return Value:
+//    false if no changes were made
+//
+bool Lowering::TryLowerAndOrToCCMP(GenTreeOp* tree, GenTree** next)
+{
+    assert(tree->OperIs(GT_AND, GT_OR));
+
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return false;
+    }
+
+    GenTree* op1 = tree->gtGetOp1();
+    GenTree* op2 = tree->gtGetOp2();
+
+    if ((op1->OperIsCmpCompare() && varTypeIsIntegralOrI(op1->gtGetOp1())) ||
+        (op2->OperIsCmpCompare() && varTypeIsIntegralOrI(op2->gtGetOp1())))
+    {
+        JITDUMP("[%06u] is a potential candidate for CCMP:\n", Compiler::dspTreeID(tree));
+        DISPTREERANGE(BlockRange(), tree);
+        JITDUMP("\n");
+    }
+
+    // Find out whether an operand is eligible to be converted to a conditional
+    // compare. It must be a normal integral relop; for example, we cannot
+    // conditionally perform a floating point comparison and there is no "ctst"
+    // instruction that would allow us to conditionally implement
+    // TEST_EQ/TEST_NE.
+    //
+    // For the other operand we can allow more arbitrary operations that set
+    // the condition flags; the final transformation into the flags def is done
+    // by TryLowerConditionToFlagsNode.
+    //
+    //
+    // On X86, a FP compare is implemented as a fallthrough, which requires two flag checks; hence,
+    // we cannot simply get a single output condition to feed into a ccmp. Might be possible to chain
+    // this, but skipping those cases for now
+    GenCondition cond1;
+    if (op2->OperIsCmpCompare() && varTypeIsIntegralOrI(op2->gtGetOp1()) && IsInvariantInRange(op2, tree) &&
+        ProducesPotentialConsumableFlagsForCCMP(op1) && TryLowerConditionToFlagsNode(tree, op1, &cond1))
```

Comment:
I think it would be preferable to get rid of `ProducedPotentialConsumableFlagsForCCMP` and add an argument to `TryLowerConditionToFlagsNode` about whether it is allowed to lower to a condition that requires multiple flags checks. Otherwise we end up having to keep `ProducedPotentialConsumableFlagsForCCMP` and `TryLowerConditionToFlagsNode` in sync.

You can use `GenConditionDesc::Get(cond).oper == GT_NONE` to check this condition in the appropriate places in `TryLowerConditionToFlagsNode`.